File: easy-projects/easy-common/src/main/java/xyz/srclab/common/base/Format.java
package xyz.srclab.common.base;

import org.apache.commons.lang3.ArrayUtils;
import org.slf4j.helpers.MessageFormatter;
import xyz.srclab.annotation.Out;

import java.text.MessageFormat;
import java.util.Locale;

public class Format {

    public static String fastFormat(String messagePattern, Object... args) {
        return FastFormat.INSTANCE.format(messagePattern, args);
    }

    public static String printfFormat(String messagePattern, Object... args) {
        return String.format(Locale.getDefault(), messagePattern, args);
    }

    public static String messageFormat(String messagePattern, Object... args) {
        return MessageFormat.format(messagePattern, args);
    }

    private static final class FastFormat {

        private static final FastFormat INSTANCE = new FastFormat();

        public String format(String pattern, Object... args) {
            processArguments(args);
            return MessageFormatter.arrayFormat(pattern, args, null).getMessage();
        }

        private void processArguments(@Out Object... args) {
            if (ArrayUtils.isEmpty(args)) {
                return;
            }
            final Object lastEntry = args[args.length - 1];
            if (lastEntry instanceof Throwable) {
                args[args.length - 1] = lastEntry.toString();
            }
        }
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/base/Format.kt
package xyz.srclab.common.base

import org.slf4j.helpers.MessageFormatter as MessageFormatterSlf4j
import java.text.MessageFormat as MessageFormatKt

/**
 * Format chars.
 *
 * @see FastFormat
 * @see PrintfFormat
 * @see MessageFormat
 */
interface Format {

    fun format(pattern: CharSequence, vararg args: Any?): String

    companion object {

        @JvmField
        val FAST_FORMAT: FastFormat = FastFormat

        @JvmField
        val PRINT_FORMAT: PrintfFormat = PrintfFormat

        @JvmField
        val MESSAGE_FORMAT: MessageFormat = MessageFormat

        @JvmStatic
        fun CharSequence.fastFormat(vararg args: Any?): String {
            return FastFormat.format(this, *args)
        }

        @JvmStatic
        fun CharSequence.printfFormat(vararg args: Any?): String {
            return PrintfFormat.format(this, *args)
        }

        @JvmStatic
        fun CharSequence.messageFormat(vararg args: Any?): String {
            return MessageFormat.format(this, *args)
        }
    }
}

/**
 * [Format] with slf4j style.
 */
object FastFormat : Format {

    override fun format(pattern: CharSequence, vararg args: Any?): String {
        processArguments(args)
        return MessageFormatterSlf4j.arrayFormat(pattern.toString(), args, null).message
    }

    private fun processArguments(array: Array<out Any?>) {
        if (array.isEmpty()) {
            return
        }

        /*
        val lastElement = args[args.size - 1]
        if (lastElement is Throwable) {
            args[args.size - 1] = lastElement.toString()
        }
         */
    }
}

/**
 * [Format] with `System.out.printf` style.
 */
object PrintfFormat : Format {

    override fun format(pattern: CharSequence, vararg args: Any?): String {
        return String.format(Default.locale, pattern.toString(), *args)
    }
}

/**
 * [Format] with `String.format` style.
 */
object MessageFormat : Format {

    override fun format(pattern: CharSequence, vararg args: Any?): String {
        return MessageFormatKt.format(pattern.toString(), *args)
    }
}